### PR TITLE
Fyle-finance-dashboard

### PR DESCRIPTION
**Pull request by sheshantsinha**

### CODE REVIEW

File: apps/users/urls.py [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+    1. Add an import:  from other_app.views import Home
+    2. Add a URL to urlpatterns:  path('', Home.as_view(), name='home')
+Including another URLconf
+    1. Import the include() function: from django.urls import include, path
+    2. Add a URL to urlpatterns:  path('blog/', include('blog.urls'))
+"""
+from django.urls import path
+
+from .views import UserProfileView
+from ..workspaces.views import UserAccountMapping
```
Comment (Shwetabhk) [inline apps/users/urls.py:19]: workspaces should be called `enterprises`

File: apps/users/urls.py [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+
+urlpatterns = [
+    path('profile/', UserProfileView.as_view()),
+    path('map_user/', UserAccountMapping.as_view())
```
Comment (Shwetabhk) [inline apps/users/urls.py:23]: If the URL is mentioned here it should be added in users/views.py
else add it in enterprises/urls.py

File: apps/users/views.py [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+
+    permission_classes = [IsAuthenticated]
+
+    def get(self, request, *args, **kwargs):
+        """
+        Get User Details
+        """
+        fyle_credentials = AuthToken.objects.get(user__user_id=request.user)
+
+        fyle_connector = FyleConnector(fyle_credentials.refresh_token)
```
Comment (Shwetabhk) [inline apps/users/views.py:18]: This will error out, it is not imported

File: apps/workspaces/apps.py [hunk 4]
```diff
@@ -0,0 +1,5 @@
+from django.apps import AppConfig
+
+
+class WorkspaceConfig(AppConfig):
+    name = 'workspaces'
```
Comment (Shwetabhk) [inline apps/workspaces/apps.py:5]: call it `enterprises`

File: apps/workspaces/models.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+User = get_user_model()
+
+
+class Enterprise(models.Model):
+    """
+    Enterprise
+    """
+    created_at = models.DateTimeField(auto_now_add=True, help_text='created at time')
+    updated_at = models.DateTimeField(auto_now=True, help_text='updated at time')
+    id = models.AutoField(primary_key=True, serialize=False)
```
Comment (Shwetabhk) [inline apps/workspaces/models.py:16]: always add id at the top

File: apps/workspaces/models.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+from django.contrib.auth import get_user_model
+
+User = get_user_model()
+
+
+class Enterprise(models.Model):
+    """
+    Enterprise
+    """
+    created_at = models.DateTimeField(auto_now_add=True, help_text='created at time')
```
Comment (Shwetabhk) [inline apps/workspaces/models.py:14]: always add timestamps at the bottom

File: apps/workspaces/models.py [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+    users = models.ManyToManyField(User, help_text='User id mapping to enterprise table')
+
+
+class Orgs(models.Model):
+    """
+    Orgs
+    """
+    created_at = models.DateTimeField(auto_now_add=True, help_text='created at time')
+    updated_at = models.DateTimeField(auto_now=True, help_text='updated at time')
+    id = models.AutoField(primary_key=True, serialize=False)
```
Comment (Shwetabhk) [inline apps/workspaces/models.py:27]: same here

Comment (Shwetabhk) [inline apps/workspaces/models.py:27]: remove `serialize=False` from everywhere

File: apps/workspaces/models.py [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+    org_name = models.TextField(blank=True, null=True, verbose_name='Org name')
+
+
+class Exports(models.Model):
+    """
+    Exports
+    """
+    created_at = models.DateTimeField(auto_now_add=True, help_text='created at time')
+    updated_at = models.DateTimeField(auto_now=True, help_text='updated at time')
+    id = models.AutoField(primary_key=True, serialize=False)
```
Comment (Shwetabhk) [inline apps/workspaces/models.py:40]: same here

File: apps/workspaces/models.py [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+    enterprise = models.ForeignKey(Enterprise, on_delete=models.deletion.PROTECT)
```
Comment (Shwetabhk) [inline apps/workspaces/models.py:41]: it should be `models.PROTECT` everywhere

File: apps/workspaces/models.py [hunk 10]
```diff
@@ -0,0 +1,44 @@
...
+
+
+class Orgs(models.Model):
+    """
+    Orgs
+    """
+    created_at = models.DateTimeField(auto_now_add=True, help_text='created at time')
+    updated_at = models.DateTimeField(auto_now=True, help_text='updated at time')
+    id = models.AutoField(primary_key=True, serialize=False)
+    enterprise = models.ForeignKey(Enterprise, on_delete=models.deletion.PROTECT)
```
Comment (Shwetabhk) [inline apps/workspaces/models.py:28]: `models.PROTECT`

File: apps/workspaces/permissions.py [hunk 11]
```diff
@@ -0,0 +1,21 @@
+from django.contrib.auth import get_user_model
+from rest_framework import permissions
+
+from apps.workspaces.models import Workspace
+
+User = get_user_model()
+
+
+class WorkspacePermissions(permissions.BasePermission):
```
Comment (Shwetabhk) [inline apps/workspaces/permissions.py:9]: EnterprisePermissions

File: apps/workspaces/permissions.py [hunk 12]
```diff
@@ -0,0 +1,21 @@
...
+    """
+    Permission check for users <> workspaces
+    """
+
+    def has_permission(self, request, view):
```
Comment (Shwetabhk) [inline apps/workspaces/permissions.py:14]: use this permission to check whether the user exists in enterprise or not

File: apps/enterprise/urls.py [hunk 13]
```diff
@@ -0,0 +1,27 @@
...
+    1. Import the include() function: from django.urls import include, path
+    2. Add a URL to urlpatterns:  path('blog/', include('blog.urls'))
+"""
+from django.urls import path, include
+
+from .views import EnterpriseView, UserAccountMapping
+
+urlpatterns = [
+    path('orgs/<int:enterprise_id>', EnterpriseView.as_view({'get': 'get_orgs'})),
+    path('exports/<int:enterprise_id>', EnterpriseView.as_view({'get': 'get_export'})),
```
[the remark this answers is not in the document]
Comment (Shwetabhk) [reply apps/enterprise/urls.py:22]: POST export should not have another endpoint, use this one

File: apps/enterprise/urls.py [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+    path('exports/', EnterpriseView.as_view({'post': 'post_export'})),
+    path('user/map_user/', UserAccountMapping.as_view()),
+    path('enterprise/', EnterpriseView.as_view({'post': 'post_enterprise'})),
+    path('orgs/', EnterpriseView.as_view({'post': 'post_org'}))
```
Comment (Shwetabhk) [inline apps/enterprise/urls.py:26]: POST orgs should not be another endpoint
it should be under 1 view

File: apps/enterprise/views.py [hunk 15]
```diff
@@ -0,0 +1,186 @@
...
+from .models import Enterprise, Orgs, Exports
+from ..users.models import User
+from .serializers import ExportSerializer, EnterpriseSerializer, OrgsSerializer
+from ..exports import gsheet, fyle
+
+User = get_user_model()
+auth_utils = AuthUtils()
+
+
+class EnterpriseView(viewsets.ViewSet):
```
Comment (Shwetabhk) [inline apps/enterprise/views.py:20]: This is wrong and not at all in line with the patter we're following in QBO
`EnterpriseView` should only have GET and POST of Enterprise
`OrgView` should have GET and POST of orgs
`Export View` should have GET and POST of exports

File: fyle-finance-dashboard-api/settings.py [hunk 16]
```diff
@@ -0,0 +1,171 @@
...
+    },
+]
+
+FYLE_REST_AUTH_SERIALIZERS = {
+    'USER_DETAILS_SERIALIZER': 'apps.users.serializers.UserSerializer'
+}
+
+REST_FRAMEWORK = {
+    'DEFAULT_PERMISSION_CLASSES': (
+        'rest_framework.permissions.IsAuthenticated',
```
Comment (Shwetabhk) [inline fyle-finance-dashboard-api/settings.py:94]: Enterprise permission is not mentioned here

File: apps/enterprise/views.py [hunk 17]
```diff
@@ -0,0 +1,186 @@
...
+    """
+    Fyle Finance Export
+    """
+
+    permission_classes = [IsAuthenticated]
```
Comment (Shwetabhk) [inline apps/enterprise/views.py:25]: this is not needed here, if the permission is mentioned in settings.py

File: apps/enterprise/views.py [hunk 18]
```diff
@@ -0,0 +1,186 @@
...
+
+    def post_export(self, request):
+
+        enterprise_id = request.data['id']
+        tpa_data_format = fyle.FormatData()
```
Comment (Shwetabhk) [inline apps/enterprise/views.py:30]: all of this logic should be not be in the views
Views should only be used for request and response

File: apps/exports/fyle.py [hunk 19]
```diff
@@ -0,0 +1,114 @@
+from fylesdk import FyleSDK
+import os
+
+class FyleTpaData:
```
Comment (Shwetabhk) [inline apps/exports/fyle.py:4]: FyleTPAData name is no right
Please use the same convention that we use in QBO

File: apps/exports/fyle.py [hunk 20]
```diff
@@ -0,0 +1,114 @@
...
+    def __init__(self, refresh_token):
+        client_id = os.environ['FYLE_CLIENT_ID']
```
Comment (Shwetabhk) [inline apps/exports/fyle.py:6]: we should not read credentials directly from the environment
environment variables should be read in django Settings file and from the settings we'll read these

File: apps/exports/fyle.py [hunk 21]
```diff
@@ -0,0 +1,114 @@
...
+        client_secret = os.environ['FYLE_CLIENT_SECRET']
+        self.conn = FyleSDK(
+            base_url='https://staging.fyle.in',
+            client_id=client_id,
+            client_secret=client_secret,
+            refresh_token=refresh_token
+        )
+
+    def fyle_tpa(self):
```
Comment (Shwetabhk) [inline apps/exports/fyle.py:15]: The name is wrong here
function names should always mention the action they are performing and should be verbs
for example `get_expenses`
Please check QBO code

File: apps/exports/fyle.py [hunk 22]
```diff
@@ -0,0 +1,114 @@
...
+        return self.conn.Expenses.get_all()
+
+
+class FormatData:
```
Comment (Shwetabhk) [inline apps/exports/fyle.py:19]: Why is  this a class?

File: apps/fyle/utils.py [hunk 23]
```diff
@@ -0,0 +1,28 @@
+from django.conf import settings
+
+from fylesdk import FyleSDK
+
+
+class FyleConnector:
```
Comment (Shwetabhk) [inline apps/fyle/utils.py:6]: This is not something I understand, if we already have FyleConnector why do we need FyleTPAData?
Please remove Fyle TPA Data and use this

File: apps/enterprise/views.py (hunk 15 above)
Comment (Shwetabhk) [inline apps/enterprise/views.py:20]: Use Generic view everywhere view sets are not needed anywhere
https://www.django-rest-framework.org/api-guide/generic-views/
Check qbo code for examples

File: apps/exports/fyle.py (hunk 22 above)
Comment (sheshantsinha) [reply apps/exports/fyle.py:19]: this is for formatting data into 2D array and mapping headers

**Pull request closed without merging**